### PR TITLE
Allow atomic swap merger to receive claim/timeout before open

### LINKED ISSUE #556: Occasional CI fail "Get and watch atomic swap lifecycle"
```
1. BnsConnection : Get and watch atomic swap lifecycle
   Expected 4 to equal 5.
       at <Jasmine>
       at Object.<anonymous> /home/travis/build/iov-one/iov-core/packages/iov-bns/src/bnsconnection.spec.ts:1376:25
       at <Jasmine>
       at fulfilled /home/travis/build/iov-one/iov-core/packages/iov-bns/build/bnsconnection.spec.js:4:58
   Expected 'open' to equal 'claimed'.
       at <Jasmine>
       at Object.<anonymous> /home/travis/build/iov-one/iov-core/packages/iov-bns/src/bnsconnection.spec.ts:1381:26
       at <Jasmine>
       at fulfilled /home/travis/build/iov-one/iov-core/packages/iov-bns/build/bnsconnection.spec.js:4:58
   Expected $['7'] = 4 to equal 3.
       at <Jasmine>
       at Object.<anonymous> /home/travis/build/iov-one/iov-core/packages/iov-bns/src/bnsconnection.spec.ts:1382:29
       at <Jasmine>
       at fulfilled /home/travis/build/iov-one/iov-core/packages/iov-bns/build/bnsconnection.spec.js:4:58
   Expected 'claimed' to equal 'open'.
       at <Jasmine>
       at Object.<anonymous> /home/travis/build/iov-one/iov-core/packages/iov-bns/src/bnsconnection.spec.ts:1383:26
       at <Jasmine>
       at fulfilled /home/travis/build/iov-one/iov-core/packages/iov-bns/build/bnsconnection.spec.js:4:58
   Expected $['7'] = 2 to equal 4.
       at <Jasmine>
       at Object.<anonymous> /home/travis/build/iov-one/iov-core/packages/iov-bns/src/bnsconnection.spec.ts:1384:29
       at <Jasmine>
       at fulfilled /home/travis/build/iov-one/iov-core/packages/iov-bns/build/bnsconnection.spec.js:4:58
   TypeError: Cannot read property 'kind' of undefined
       at <Jasmine>
       at Object.<anonymous> /home/travis/build/iov-one/iov-core/packages/iov-bns/src/bnsconnection.spec.ts:1385:20
       at <Jasmine>
       at fulfilled /home/travis/build/iov-one/iov-core/packages/iov-bns/build/bnsconnection.spec.js:4:58
       at <Jasmine>
       at process._tickCallback internal/process/next_tick.js:189:7
```

https://travis-ci.com/iov-one/iov-core/jobs/160267270

### PR DESCRIPTION
**Pull request by webmaster128**
Closes #556

### CODE REVIEW

File: packages/iov-bcp-types/src/atomicswapmerger.ts [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    };
+  } else {
+    return {
+      kind: SwapState.Expired,
+      data: swap.data,
+    };
+  }
+}
+
+export class AtomicSwapMerger {
```
Comment (ethanfrey) [inline packages/iov-bcp-types/src/atomicswapmerger.ts]: Nice, so this takes a series of updates/transactions and locally materializes a state from it? I like it

File: packages/iov-bcp-types/src/transactions.ts [hunk 2]
```diff
@@ -209,7 +209,7 @@ export interface SendTransaction extends UnsignedTransaction {
 
 export interface SwapOfferTransaction extends UnsignedTransaction {
   readonly kind: "bcp/swap_offer";
-  readonly amount: ReadonlyArray<Amount>;
+  readonly amounts: ReadonlyArray<Amount>;
```
Comment (ethanfrey) [inline packages/iov-bcp-types/src/transactions.ts]: yes, better name

File: packages/iov-bcp-types/src/atomicswapmerger.ts (hunk 1 above)
Comment (webmaster128) [reply packages/iov-bcp-types/src/atomicswapmerger.ts]: Jap, more or less what you did inline before but extracted in a class dedicated to that job.

**Comment by webmaster128** [comment]
By the way, this is the point at which `@iov/bcp-types` should be renamed to `@iov/bcp`. For your mental modal assume this renaming was already performed.

**Pull request merged**